Adds API info to RFC (#9)
elorzafe/amplify-types: RFC.md

Try out the new `storage` types here: https://www.typescriptlang.org/play#gist/292f4e24178bfca5881aa20961b930dc

> # `API` category Changes

Reviewer: ```suggestion
# `API` Category Changes
Amplify is proposing the following changes for the `API` category.
```

(for top level sections)

## First param is an object with named parameters

> 

Reviewer: ```suggestion
To improve the readability of our APIs we will be introducing an object parameter to capture request parameters.
```

**Amplify v5 (`aws-amplify@5`)**

```typescript
const apiName = 'MyApiName';

> const path = '/path';

Reviewer: Nit: IMO these should just be inline (e.g. `API.get('MyApiName', '/path', ...)`)

Author: I copy/pasted this from our docs actually. Is just to show is better the name approach so is easy to do it inline

const myInit = {
  headers: {}, // OPTIONAL
  response: true, // OPTIONAL (return the entire Axios response object instead of only response.data)
  queryStringParameters: {
    name: 'param' // OPTIONAL
  }
};

API.get(apiName, path, myInit)
  .then((response) => {
    // Add your code here
  })
  .catch((error) => {
    console.log(error.response);
  });
```

> 

Reviewer: ```suggestion
**Proposed Amplify v6 (`aws-amplify@6`)**
```

```typescript
API.get({
    apiName: 'MyApi',
    path: '/items',

>     authMode: 'AWS_IAM'

Reviewer: ```suggestion
  apiName: 'MyApi',
  path: '/items',
  authMode: 'AWS_IAM'
```

> }, myInit)

Reviewer: Doesn't need to happen now, but why was `myInit` excluded from the param object? We should try to be consistent between category APIs one way or the other.

Author: You mean having two arguments?

Reviewer: Yeah in this PR it has two arguments but I was wondering if it should just be one object param

>   .then(result => {

Reviewer: ```suggestion
  .then((result) => {
```

Inconsistent parens

    // Add your code here
  }).catch((error) => {
    console.log(error.response);
  });
```

> ## Adding TypeScript generics to request body and request response

Reviewer: ```suggestion
## Adding TypeScript generics to request body and response
To improve developer experience and permit more strict typing we will be adding generic support to our `API` category APIs.
```

**Amplify v5 (`aws-amplify@5`)**

> Not possible to use generics for request body or response body
> 

Reviewer: ```suggestion
Amplify v5 does not support using generics for the request body or response.
```

**Proposed Amplify v6 (`aws-amplify@6`)**

```typescript
type MyApiResponse = { firstName: string, lastName: string };

API.get<MyApiResponse>({
    apiName: 'MyApi',
    path: '/getName'
}).then(result => {
    console.log(`The name is ${result.body.firstName} ${result.body.lastName}`)
});

API.put<string, { data: Array<number> }>({
    apiName: '',
    path: '/',
    authMode: "API_KEY"
}, {
    headers: {
        "Content-type": "text/plain",
    },
    body: "this is my content"

> }).then(result => {
>     result.body.data.forEach(value => console.log(value));

Reviewer: ```suggestion
}).then((result) => {
    result.body.data.forEach((value) => console.log(value));
```

(consistent parens)

});
```

> ## GraphQL operations are splitted between query, mutation and subscription

Reviewer: ```suggestion
## GraphQL operations have been split into query, mutation, and subscription
To better capture customer intent and simplify API types we will split up the `graphql` API into individual APIs for queries, mutations, and subscriptions.
```

**Amplify v5 (`aws-amplify@5`)**

```typescript

> import { API } from "aws-amplify";
> import * as mutations from './graphql/mutations';
> import { GraphQLQuery, GraphQLSubscription } from '@aws-amplify/api';
> import { CreateTodoInput, CreateTodoMutation, OnCreateTodoSubscription } from './API';
> import * as subscriptions from './graphql/subscriptions';
> 

Reviewer: ```suggestion
```

IMO we should omit imports from samples.

Author: Yeah, it was to make explicit the thing is annoying lol

const todoDetails: CreateTodoInput = {
  name: 'Todo 1',
  description: 'Learn AWS AppSync'
};

const newTodo = await API.graphql<GraphQLQuery<CreateTodoMutation>>({ 
  query: mutations.createTodo, 
  variables: { input: todoDetails }
});

> const sub = API.graphql<GraphQLSubscription<OnCreateTodoSubscription>>(

Reviewer: ```suggestion
const subscription = API.graphql<GraphQLSubscription<OnCreateTodoSubscription>>(
```

  graphqlOperation(subscriptions.onCreateTodo)
).subscribe({
  next: ({ provider, value }) => console.log({ provider, value }),
  error: (error) => console.warn(error)
});
```
**Proposed Amplify v6 (`aws-amplify@6`)**

```typescript
type MyQueryType = {

>     result: {
>         id: string,
>         name: string,
>         description: string
>     }

Reviewer: ```suggestion
  result: {
    id: string,
    name: string,
    description: string
  }
```

}

> API.graphqlQuery<MyQueryType>({

Reviewer: It seems like the "query" example is missing from v5 sample above, we should make sure the v5 & v6 samples capture identical tasks to make comparison easier.

>     document: `query getTodo...`,

Reviewer: ```suggestion
  document: `query getTodo...`,
```

>     document: `query getTodo...`,

Reviewer: Any reason to not using the codegen'd docs here? Don't want to imply that customers will need to write their own documents.

Author: Good call, it was more about telling the customer they could also use their own queries (not depending on CLI) I didnt like the idea of having too many things that they might ask where is this coming from

})
.then(result => {

>     console.log(`Todo : ${result.data?.id}: ${result.data?.name} (${result.data?.description})`);

Reviewer: ```suggestion
.then((result) => {
  console.log(`Todo : ${result.data?.id}: ${result.data?.name} (${result.data?.description})`);
```

});

type MyMutationType = {
    variables: {
        id: number,
        name: string,
        description: string
    },
    result: {
      id: number, 
      name: string, 
      description: string

>     }

Reviewer: ```suggestion
  variables: {
    id: number,
    name: string,
    description: string
  },
  result: {
    id: number, 
    name: string, 
    description: string
  }
```

}

API.graphqlMutation<MyMutationType>({

>     document: `mutation createTodo....`,

Reviewer: Ditto about using existing document.

>     document: `mutation createTodo....`,
>     variables: {
>         id: 123,
>         name: 'My Todo',
>         description: 'This is a todo'
>     }

Reviewer: ```suggestion
  document: `mutation createTodo....`,
  variables: {
    id: 123,
    name: 'My Todo',
    description: 'This is a todo'
  }
```

})
.then(result => {

>     console.log(`Todo : ${result.data?.id}: ${result.data?.name} (${result.data?.description})`);

Reviewer: ```suggestion
.then((result) => {
  console.log(`Todo : ${result.data?.id}: ${result.data?.name} (${result.data?.description})`);
```

});

> API.graphqlSubscription<MyQueryType>({ document: `subscription OnCreateTodo...`})

Reviewer: Ditto about using document

.subscribe({

>     next: (result) => console.log(`Todo info: ${result.data?.id}: ${result.data?.name} (${result.data?.description})`),

Reviewer: ```suggestion
  next: (result) => console.log(`Todo info: ${result.data?.id}: ${result.data?.name} (${result.data?.description})`),
```

});
```

## Errors can be narrowed down 

**Amplify v5 (`aws-amplify@5`)**

> Errors cannot be narrowed down

Reviewer: ```suggestion
Amplify v5 does not support narrowing down errors.
```

**Proposed Amplify v6 (`aws-amplify@6`)**

```typescript
import { HTTPError, NetworkError, BlockedError, CancelledError}

> 

Reviewer: ```suggestion
```

API.get({

>     apiName: 'myApi',
>     path: '/'

Reviewer: ```suggestion
  apiName: 'myApi',
  path: '/'
```

}).then(result => {

>     // do something with result

Reviewer: ```suggestion
}).then((result) => {
  // do something with result
```

}).catch((err: unknown) => {

>     if (err instanceof NetworkError) {
>         // Maybe I want to retry
>     } else if (err instanceof HTTPError) {
>         // Check if something is going with my request parameters
>     } else if (err instanceof CancelledError) {
>         // this is fine I trigger the cancellation
>     } else if (err instanceof BlockedError) {
>         // This is CORS stuff
>     } else {
>         // other error
>     }

Reviewer: ```suggestion
  if (err instanceof NetworkError) {
    // Consider retrying
  } else if (err instanceof HTTPError) {
    // Check request parameters for mistakes
  } else if (err instanceof CancelledError) {
    // Request was cancelled
  } else if (err instanceof BlockedError) {
    // CORS related error
  } else {
    // Other error
  }
```

});
```

> Try out the new types here: TODO Playground Link

Reviewer: Could you get this updated? Storage example here: https://www.typescriptlang.org/play#gist/292f4e24178bfca5881aa20961b930dc

Author: I miss that, I thought it was there, updating now
